test/int/adapter.test.ts

```diff
@@ -21,7 +21,7 @@ suite('Chrome Debug Adapter etc', () => {
     });
 
     teardown(() => {
-        return testSetup.teardown();
+        testSetup.teardown();
```

Comment:
Yes. Sorry, I made this change really quickly to get the tests to stop hanging and forgot the details. The problem is that this is the function that gets called in mocha for the setup/teardown hooks:
![image](https://user-images.githubusercontent.com/16889261/51555031-8bd61180-1e2b-11e9-9f83-a1692b73096e.png)

When we return a promise in the setup/teardown, mocha executes the first if block. It attempts to pass a resolve function to the pending promise with a call to its own "done()" method. However, we never take in a resolve/reject function, so "done()" never gets called, and mocha interprets this as a test not cleaning up after itself properly, so it hangs there. 

The reason this isn't the case with the setup function is because the setup function (eventually) returns this promise which is in the debugClient of vscode-debugadapter-testsupport:
![image](https://user-images.githubusercontent.com/16889261/51555069-a14b3b80-1e2b-11e9-8210-7871830ee15d.png)

The start method calls resolve, whereas the stop method (also in the debugClient of vscode-debugadapter-testsupport), which is what eventually gets called in the teardown method, doesn't.
![image](https://user-images.githubusercontent.com/16889261/51555094-ab6d3a00-1e2b-11e9-9306-fc870f99388e.png)

If we don't return a promise, the return value from the teardown method is undefined, so mocha executes the second part of the if block, and the done method gets called and doesn't hang. 

After looking at it again I'm realizing a better way to do it would probably be to change the stop() method to this:
		`return new Promise((resolve, reject) => {
			this.disconnectRequest().then(() => {
				this.stopAdapter();
			}).catch(() => {
				this.stopAdapter();
				reject();
			});
			
			resolve();
		});`

Reply:
I don't really follow, we are returning a promise, and the should be resolved. If it isn't being resolved then that's wrong and we should figure out why that's happening, not ignore the promise.

Reply:
Your suggested change to `stop` would cause the promise to be resolved all the time which doesn't help the problem of `disconnectRequest` not being resolved.

Reply:
The disconnectRequest promise does get resolved, even when returning the promise, but mocha's `done()` method never gets called, so after setting up the first test, executing the first test, and (successfully) tearing down the first test, it never moves on beyond that point.

Sorry, you're right about not wanting to resolve all the time. 
```
return new Promise((resolve, reject) => {
    this.disconnectRequest().then(() => {
        this.stopAdapter();
        resolve();
    }).catch(() => {
        this.stopAdapter();
        reject();
    });
```
Would this be preferable?

Reply:
> but mocha's done() method never gets called,

If disconnectRequest is resolve, then I don't understand where it's getting stuck.

I think your new code is exactly equivalent to what what was there previously, except that it will be rejected if disconnect fails, which I don't know whether we want that or not.

Reply:
Mocha gets stuck because mocha doesn't proceed without its `done()` method being called. For example, mocha hangs with this code:
![image](https://user-images.githubusercontent.com/16889261/51563035-1b85bb00-1e40-11e9-9a82-53cbfc31a45f.png)

But not with this code:
![image](https://user-images.githubusercontent.com/16889261/51562899-c47fe600-1e3f-11e9-8415-739d43fdc79b.png)

Mocha requires its `done()` method to be called in order for it to move on to the next test (the red circled methods):
![image](https://user-images.githubusercontent.com/16889261/51563089-46700f00-1e40-11e9-83dc-20aa769dce79.png)

Reply:
I still don't understand, that must be because the teardown promise isn't resolved. In the suggested code you just return an automatically resolved promise regardless of what teardown does. The mocha code just shows that `done()` will only run if the promise is resolved.

Reply:
Hey let me know if you want to come down and debug it together, promises can be confusing

Reply:
Thank you for your help!